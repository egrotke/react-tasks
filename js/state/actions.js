export const ADD_TASK = 'ADD_TASK';
export const DELETE_TASK = 'DELETE_TASK';
export const SAVE_TASKS = 'SAVE_TASKS';
export const ADD_TASKS_FROM_API = 'ADD_TASKS_FROM_API';
export const SHOW_SAVED_MESSAGE = 'SHOW_SAVED_MESSAGE';
export const HIDE_SAVED_MESSAGE = 'HIDE_SAVED_MESSAGE';
export const SHOW_ADD_MODAL = 'SHOW_ADD_MODAL';

