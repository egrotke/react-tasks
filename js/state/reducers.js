import uid from 'uid';
import {
   ADD_TASK,
   DELETE_TASK,
   ADD_TASKS_FROM_API,
   SHOW_SAVED_MESSAGE,
   HIDE_SAVED_MESSAGE,
   SHOW_ADD_MODAL
} from './actions';

const addTask = (state, action) => {
   if (!action.payload) {
      return state;
   }
   const { tasks } = state;
   const task = { id: uid(tasks.length), title: action.payload };
   return Object.assign({}, state, { tasks: [task, ...tasks] });
};
const deleteTask = (state, action) => {
   const { tasks } = state;
   return Object.assign({}, state, {
      tasks: tasks.filter(task => task.id !== action.payload)
   });
};
const getTasks = (state, action) => {
   const { tasks } = state;
   return Object.assign({}, state, tasks, action.payload);
};
const showAddModal = (state, action) =>
   Object.assign({}, state, { showAddTaskModal: action.payload });

const showSavedMessage = state => Object.assign({}, state, { showAlert: true });

const hideSavedMessage = state =>
   Object.assign({}, state, { showAlert: false });

const rootReducer = (state, action) => {
   switch (action.type) {
      case ADD_TASK:
         return addTask(state, action);
      case DELETE_TASK:
         return deleteTask(state, action);
      case ADD_TASKS_FROM_API:
         return getTasks(state, action);
      case SHOW_SAVED_MESSAGE:
         return showSavedMessage(state);
      case HIDE_SAVED_MESSAGE:
         return hideSavedMessage(state);
      case SHOW_ADD_MODAL:
         return showAddModal(state, action);
      default:
         return state;
   }
};

export default rootReducer;
