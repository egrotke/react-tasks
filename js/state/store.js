import { createStore, compose, applyMiddleware } from 'redux'; // add applyMiddleware
import thunk from 'redux-thunk'; // import
import rootReducer from './reducers';

const initialState = {
   showAlert: false,
   showAddTaskModal: false,
   tasks: []
};
const store = createStore(
   rootReducer,
   initialState,
   compose(
      applyMiddleware(thunk), // middleware
      typeof window === 'object' &&
         typeof window.devToolsExtension !== 'undefined'
         ? window.devToolsExtension()
         : f => f
   )
);

export default store;
