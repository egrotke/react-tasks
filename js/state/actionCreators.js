import axios from 'axios';
import store from './store';

import {
   ADD_TASK,
   DELETE_TASK,
   SHOW_SAVED_MESSAGE,
   HIDE_SAVED_MESSAGE,
   ADD_TASKS_FROM_API,
   SHOW_ADD_MODAL
} from './actions';

export function addTask(title) {
   return { type: ADD_TASK, payload: title };
}
export function deleteTask(id) {
   return { type: DELETE_TASK, payload: id };
}
export function addTasksFromAPI(tasks) {
   return { type: ADD_TASKS_FROM_API, payload: tasks };
}
export function showAddModal(showOrHide) {
   return { type: SHOW_ADD_MODAL, payload: showOrHide };
}

export function showSavedMessage() {
   return { type: SHOW_SAVED_MESSAGE };
}
export function saveTasks() {
   const { tasks } = store.getState();
   return dispatch => {
      axios
         .post('http://cfassignment.herokuapp.com/egrotke/tasks', {
            tasks
         })
         .then(() => {
            dispatch(showSavedMessage());
         })
         .catch(error => {
            console.error('axios error', error); // eslint-disable-line no-console
         });
   };
}
export function hideSavedMessage() {
   return { type: HIDE_SAVED_MESSAGE };
}
// export function fetchTasks() {
//    return dispatch => {
//       axios
//          .get('http://cfassignment.herokuapp.com/egrotke/tasks')
//          .then(response => {
//             dispatch(getTasks(response.data));
//          })
//          .catch(error => {
//             console.error('axios error', error); // eslint-disable-line no-console
//          });
//    };
// }
