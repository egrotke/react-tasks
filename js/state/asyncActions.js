// @flow

import axios from 'axios';
import { addTasksFromAPI } from './actionCreators';

export default function fetchTasks() {
   return (dispatch: Function) => {
      axios
         .get('http://cfassignment.herokuapp.com/egrotke/tasks')
         .then(response => {
            dispatch(addTasksFromAPI(response.data));
         })
         .catch(error => {
            console.error('axios error', error); // eslint-disable-line no-console
         });
   };
}
