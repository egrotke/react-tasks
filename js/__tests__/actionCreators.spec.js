// @flow

import { addTask, deleteTask } from '../state/actionCreators';
// import moxios from 'moxios';
// import fetchTasks from '../state/asyncActions';

test('addTask', () => {
  expect(addTask('Laundry day')).toMatchSnapshot();
});

test('deleteTask', () => {
  expect(deleteTask()).toMatchSnapshot();
});

// test('fetchTasks', (done: Function) => {
//   const dispatchMock = jest.fn();
//   moxios.withMock(() => {
//     fetchTasks()(dispatchMock);
//     moxios.wait(() => {
//       const request = moxios.requests.mostRecent();
//       request
//         .respondWith({
//           status: 200,
//           response: ''
//         })
//         .then(() => {
//           expect(request.url).toEqual(
//             `http://cfassignment.herokuapp.com/egrotke/tasks`
//           );
//           expect(dispatchMock).toBeCalledWith(fetchTasks());
//           done();
//         });
//     });
//   });
// });
