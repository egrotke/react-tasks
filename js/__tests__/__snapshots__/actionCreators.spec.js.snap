// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTask 1`] = `
Object {
  "payload": "Laundry day",
  "type": "ADD_TASK",
}
`;

exports[`deleteTask 1`] = `
Object {
  "payload": undefined,
  "type": "DELETE_TASK",
}
`;
